File: FileSystemManipulation/FileSystemManipulator/FileSystemManipulatorConfiguration.m
#import "FileSystemManipulatorConfiguration.h"
#import "FileSystemManipulator.h"

@interface FileSystemManipulatorConfiguration ()
@property (nonatomic, strong) NSMutableDictionary<NSString *, NSString *> *filesToCopy;
@end

@implementation FileSystemManipulatorConfiguration

- (instancetype)init {
    if (self = [super init]) {
        _filesToCopy = [NSMutableDictionary dictionary];
    }
    return self;
}

#pragma mark - Public

+ (instancetype)configurationFromURL:(NSURL *)URL {
    NSMutableDictionary<NSString *, NSString *> *filesToCopy = [NSMutableDictionary dictionaryWithContentsOfURL:URL];

    if (filesToCopy) {
        FileSystemManipulatorConfiguration *configuration = [[FileSystemManipulatorConfiguration alloc] init];
        configuration->_filesToCopy = filesToCopy;
        return configuration;
    } else {
        return nil;
    }
}

- (NSDictionary<NSString *,NSString *> *)generateLaunchEnvironment {
    NSURL *configurationURL = [self writeConfigurationToFile];

    NSBundle *manipulatorBundle = [NSBundle bundleForClass:[FileSystemManipulatorConfiguration class]];
    NSString *manipulatorPath = manipulatorBundle.executablePath;

    return @{
             kFileSystemManipulatorConfigurationPathEnvironmentKey: configurationURL.path ?: @"",
             @"DYLD_INSERT_LIBRARIES": manipulatorPath ?: @""
             };
}

- (void)copyFileAtURL:(NSURL *)URL toRelativePath:(NSString *)relativeDestinationPath {
    self.filesToCopy[URL.path] = relativeDestinationPath;
}

- (void)enumerateManipulationsWithBlock:(void (^)(NSURL *sourceURL, NSString *relativeDestinationPath))block {
    [self.filesToCopy enumerateKeysAndObjectsUsingBlock:^(NSString *sourcePath, NSString *relativeDestinationPath, BOOL *stop) {
        block([NSURL fileURLWithPath:sourcePath], relativeDestinationPath);
    }];
}

#pragma mark - Private

- (NSURL *)writeConfigurationToFile {
    NSURL *configurationURL = [[NSURL fileURLWithPath:NSHomeDirectory()] URLByAppendingPathComponent:@"config"];
    [self.filesToCopy writeToURL:configurationURL atomically:YES];
    return configurationURL;
}

@end

File: Swift/FileSystemManipulation/FileSystemManipulator/FileSystemManipulatorConfiguration.swift
import Foundation

/// This class allows setting up a list of commands that the FileSystemManipulator
/// framework will carry out when it is loaded into a process. The idea is to create
/// an instance in the test process, write it to a file and pass that path to the app
/// process as an environment variable under kFileSystemManipulatorConfigurationPathEnvironmentKey
public class FileSystemManipulatorConfiguration {
    public typealias Manipulation = (source: NSURL, relativeDestination: String)

    private var filesToCopy: [String : String] = [:]

    public init() {
    }

    public init?(URL: NSURL) {
        guard let loadedFilesToCopy = NSDictionary(contentsOfURL: URL) as? [String : String] else {
            return nil
        }

        filesToCopy = loadedFilesToCopy
    }

    public func generateLaunchEnvironment() -> [String : String] {
        let configurationURL = writeConfigurationToFile()

        let manipulatorBundle = NSBundle(forClass: self.dynamicType)
        let manipulatorPath = manipulatorBundle.executablePath

        return [
            FileSystemManipulatorProcessor.ConfigurationPathEnvironmentKey: configurationURL?.path ?? "",
            "DYLD_INSERT_LIBRARIES": manipulatorPath ?? ""
        ]
    }

    public func copyFileAtURL(URL: NSURL, toRelativePath relativeDestinationPath: String) {
        filesToCopy[URL.path!] = relativeDestinationPath
    }

    public var manipulations: AnySequence<Manipulation> {
        get {
            return AnySequence(filesToCopy.map({ (NSURL(fileURLWithPath: $0.0), $0.1) }))
        }
    }

    private func writeConfigurationToFile() -> NSURL? {
        let configurationURL = NSURL(fileURLWithPath: NSHomeDirectory()).URLByAppendingPathComponent("config")
        return ((filesToCopy as NSDictionary).writeToURL(configurationURL, atomically: true) == true) ? configurationURL : nil
    }
}
